File: python/src/problems201_300/count_complete_tree_nodes.py
from abc import ABC, abstractmethod
from collections import deque
from typing import override

from src.concepts.tree_node import TreeNode


class Problem(ABC):
    @abstractmethod
    def count_nodes(self, root: TreeNode | None) -> int:
        pass


class Default(Problem):
    @override
    def count_nodes(self, root: TreeNode | None) -> int:
        if root is None:
            return 0
        count: int = 0
        queue: deque[TreeNode] = deque([root])
        while queue:
            node: TreeNode = queue.popleft()
            count += 1
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)
        return count


SOLUTIONS: list[Problem] = [Default()]

File: javascript/src/problems201_300/count-complete-tree-nodes.js
class Problem {
  countNodes(root) {
    throw new Error(`Not implemented (${root}).`);
  }
}

class Default extends Problem {
  countNodes(root) {
    if (!root) {
      return 0;
    }
    let count = 0;
    const queue = [];
    queue.push(root);
    while (queue.length > 0) {
      const node = queue.shift();
      count++;
      if (node.left) {
        queue.push(node.left);
      }
      if (node.right) {
        queue.push(node.right);
      }
    }
    return count;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
